Add timestamps to Subscription

gbirke opened the pull request:
Add createdAt and updatedAt fields with a trait. The fields will be
set automatically when the entity is saved.

wmde-manicki commented:
Sorry if I am asking something stupid but it is not clear for me how would it know to put the `createdAt` timestamp into `timestamp` field in the database?
I guess you would also like to use it in `Inquiry` and `Donation` entity classes, no?

gbirke commented:
The trait adds the fields to the entity, the database schema will be changed accordingly.
As to how the field values are set for each entity, this is done with Doctrine events, that set the value before creating and updating. The trait also takes care of that.

The fields of Inquiry and Donation need similar functionality, but can't use the traits since they have different field names. Should I put the event handing stuff in there too so all the entities have automatic time stamps?

wmde-manicki commented:
> The trait adds the fields to the entity, the database schema will be changed accordingly.

ah, right. I was still considering those entities go to the "good" old Request table. That's obviously no longer the case. Sorry for that, I'll try to more focused before writing things :)

> The fields of Inquiry and Donation need similar functionality, but can't use the traits since they have different field names. Should I put the event handing stuff in there too so all the entities have automatic time stamps?

Well, it's only my opinion, but I think those should be added to the other entities as well, at least creation timestamp. Otherwise that would lead to situation that user of entity class must remember that when creating a new entity of one type the timestamp will be generated on save, but for others it must be set "manually" before storing the entity. That kind of confusion wouldn't be nice.

Update timestamp have not been stored for "old" entities yet but it maybe it makes sense to have those for donations and inqueries as well? "Old" app wouldn't use them but maybe it makes sense in the new app, what do you FR666 guys think? I couldn't see those subscription timestamps being used in https://github.com/wmde/FundraisingFrontend (but I didn't search very hard, though), so I am not sure what's the intended usage of it. Adding update timestamps to Donation and/or Inquiry just for "compatibility" with Subscription does not sound that good too.

That said, I think those potential further changes do not need to be added in the pull request but the release having these timestamps for subscriptions should also include other modified entities.

Not related: aah, and why it's called Inquiry not the Membership?

gbirke commented:
Added the timestamp behavior to `Donation` and `Inquiry`, fixed the Factory to actually use the events that set the timestamps automatically.

gbirke commented:
@JeroenDeDauw Are you ok with using the doctrine extensions library? We we feared that it might be a bit too "magicky" for your taste.

wmde-manicki commented:
@gbirke you could now change title of this pull request as this is no longer about `Subscription` only. Plus commit message of https://github.com/wmde/FundraisingStore/commit/5151aaada956deadbb5d93add94bc33dd2a8ef70 could not be mentioning `Inquiry` as it has been renamed.

And what's about those update timestamps in `Donation` and `Membership`? Are they are going to be used somewhere in the FunFun code? Asking as they haven't seemed needed in the old codebase afair.

gbirke commented:
@wmde-manicki I changed the commit message. At the moment, the update timestamp of `Donation` and both timestamps of `Membership` are not used. I might have misread your comment and added them for consistency. If @JeroenDeDauw tells me that this is a case of YAGNI, I will remove them.

wmde-manicki commented:
thanks for polishing those descriptions!
And sorry for not being clear: my argument was that adding things that are not used just for consistency is a YAGNI. Maybe not very harmful in this case as this will be just updating the timestamp which is probably never read.

JeroenDeDauw commented:
The PR description states what it does but not why this is needed. If this stuff is being added "because it might be useful somewhere", then I think it is better not done.

> Are you ok with using the doctrine extensions library? We we feared that it might be a bit too "magicky" for your taste. 

@gbirke I've not looked at it in detail, however can say this: while libraries and frameworks do certain things for you that you do not need to do yourself, they are most certainly not cost free. Each one is yet another thing people need to be aware off, something with a different type of interface, a different release cycle and third party code that is harder to debug and modify. Do you think that when holding these kind of costs into account, it is easier for us to use this library rather than to create the required functionality ourselves? (Note that this likely will be much simpler than creating a copy of said library.)

gbirke commented:
We need the timestamps for the flood mitigation feature (https://phabricator.wikimedia.org/T123125), which is already in the old application. This is the last feature missing from completing the "Add subscription" use case.
The Donation timestamps are used in the export scripts somewhere.

In this case I'm convinced that using a library that's almost as popular as Doctrine itself and used by many other projects, is less confusing and more thoroughly tested than we could do it. We would really end up reproducing the library code if we do it ourselves.

wmde-manicki commented:
> We need the timestamps for the flood mitigation feature (https://phabricator.wikimedia.org/T123125), which is already in the old application. This is the last feature missing from completing the "Add subscription" use case.

Haven't seen the code but I get it and it's fine as a reason to have those for subscriptions. What I have been mostly kind of complaining here about is:
>  The Donation timestamps are used in the export scripts somewhere.

This is valid also for membership **but** as mentioned before only creation timestamps. `updatedAt` will be something completely new and I haven't seen a usage for it yet.

To be clear: I am not against that change in its current form. Just thought there could be some more said why add those things etc.
If you think work on this is now done, just please write it and this will be merged.

gbirke commented:
The `TimestampableEntity` trait is a kind of "Package Deal" which adds both fields. At the moment I see no use for the updatedAt field. So if needed, I can remove the magic trait and explicitly add the `createdAt` field. This would also make the field definitions more explicit.

wmde-manicki commented:
Oh, I just noticed https://github.com/wmde/FundraisingStore/commit/af6c791e76d3640ad7ec6c2b2fd9b72e342e025c. I like better it than the previous approach~

Few things that stopped me from merging the PR:
 - shouldn't the change now consist of single commit? Basically https://github.com/wmde/FundraisingStore/commit/af6c791e76d3640ad7ec6c2b2fd9b72e342e025c + Factory change + README, composer.json and possible other minor things in one nice commit?
 - shouldn't `Donation` be also updated accordingly?
 - `Membership` doesn't need to get a new `createdAt` field, the old good `timestamp` column could be reused (ie. the field of entity class should be probably renamed by the `@ORM\Column(name="timestamp")` could be used same as with `Donation`'s `dt_new`)

gbirke commented:
Squashed commits, updated README and commit msg, included suggestions by @wmde-manicki .

wmde-manicki commented on src/Entities/Donation.php:
```diff
@@ -145,6 +147,14 @@ class Donation {
+	 * @Gedmo\Timestampable(on="update")
+	 * @ORM\Column(name="updatedAt", type="datetime", nullable=true)
+	 */
+	private $updatedAt;
```
so this updateAt will not be used too, right?

gbirke replied:
aaah, that slipped through. Ok, updated PR.
